app/src/main/java/de/rampro/activitydiary/db/ActivityDiaryContentProvider.java

```diff
@@ -434,4 +437,77 @@ public int update(@NonNull Uri uri, @NonNull ContentValues values, @Nullable Str
+        String querySelection = " ";
+        SimpleDateFormat simpleDateFormat = new SimpleDateFormat(dateFormat);
+        Calendar searchedValueCal = Calendar.getInstance();
+        Calendar startValueCal = Calendar.getInstance();
+        Calendar endValueCal = Calendar.getInstance();
+
+        //Gets all start, end dates and also id of activities
+        try {
+            Cursor allRowsStart =  mOpenHelper.getReadableDatabase().rawQuery(
+                    "SELECT " + ActivityDiaryContract.Diary._ID + ", "
+                            + ActivityDiaryContract.Diary.START +", "
+                            + ActivityDiaryContract.Diary.END
+                            + " FROM " + ActivityDiaryContract.Diary.TABLE_NAME, null
+            );
+
+            if (allRowsStart.moveToFirst()) {
+                String[] columnNames = allRowsStart.getColumnNames();
+                do {
+                    String id = null, start = null, end = null;
+                    for (String name : columnNames) {
+                        if (name.equals(ActivityDiaryContract.Diary.START))
+                            start = simpleDateFormat.format(allRowsStart.getLong(allRowsStart.getColumnIndex(name)));
+                        if (name.equals(ActivityDiaryContract.Diary.END))
+                            end = allRowsStart.getLong(allRowsStart.getColumnIndex(name)) == 0 ? "0" : simpleDateFormat.format(allRowsStart.getLong(allRowsStart.getColumnIndex(name)));
+                        if (name.equals(ActivityDiaryContract.Diary._ID))
+                            id = (allRowsStart.getString(allRowsStart.getColumnIndex(name)));
+                    }
+                    //System.out.println("id: " + id + ", start: " + start + ", end: " + end);
+
+                    //Values (date, start, end) are set to specific calendar - needed for 'between' searching
+                    // (ignore values that are equals to '0' -> if end date have '0' value it means that it has not set end date
+                    // what means that this activity is still in progress)
+                    try {
+                        if (date != null)
+                            searchedValueCal.setTime(simpleDateFormat.parse(date));
+                        if (!start.equals("0") && start != null)
+                            startValueCal.setTime(simpleDateFormat.parse(start));
+                        if (end != null && !end.equals("0"))
+                            endValueCal.setTime(simpleDateFormat.parse(end));
+                    }catch (ParseException e){
+                        e.printStackTrace();
+                    }
+
+                    //here it is creating query from IDs of those activities that are matching with searched date
+                    if ((searchedValueCal.after(startValueCal) && searchedValueCal.before(endValueCal) || searchedValueCal.equals(startValueCal) || searchedValueCal.equals(endValueCal))) {
```

Comment:
Reason why I am comparing dates in Date format and not long ... is because
for example date: 10.03.2018 (233 seconds) has different long value as 10.03.2018 (without time)
So boolean if value are equals or in some cases it should not fit the range.. that's why I "cut" the time by converting it into Date (without time)

Reply:
those string compares are quite some computation effort. You might want to check out
```
Calendar c = new GregorianCalendar(2013,0,31);
c.getTimeInMillis()
```
or even somethign like
```
DateFormat.parse(...)
```
to get the date as long. If you then do the same with 24 h more you have the start and end time of that day, with this you should be able to do some intersections with START and END in the database (something like (d > START and d < END) OR (d+1>START and d+1 < END) or (d < START and d+1 > END)